runtime/lua/vim/project/_context.lua

```diff
@@ -0,0 +1,195 @@
+  return context_data
+end
+
+--- Deletes a key from the project context data or clears the entire context data if no key is provided.
+--- @param key string|nil The key to delete from the context data. If nil, it clears the entire context data.
+--- @return nil
+function M.del_context(key)
+  vim.validate('key', key, function(v)
+    return v == nil or (type(v) == 'string' and v ~= '')
+  end)
+
+  local context_data = M.get_context()
+  if key then
+    if context_data and context_data[key] then
+      context_data[key] = nil
+    end
+  else
+    context_data = {}
+  end
+
+  write_context_file(context_data)
+end
+
+--- Checks if a file or directory is excluded based on the provided exclude patterns.
+local function is_excluded(exclude_patterns, path)
+  --- TODO: implement file/folder exclusion
+  return false
+end
+
+--- Traverses the directory recursively and collects files and directories based on the provided options.
+--- @param opts table Options for traversal
+--- @param dir string The directory to start traversing from
+--- @param depth number The current depth in the directory tree
+--- @param files table The table to collect files and directories
+--- @return table A table containing the collected files and directories
+local function traverse_directory(opts, dir, depth, files)
+  -- TODO: upstream directory traversal from planery.scandir to core as part of vim.fs ?
+  if not files then
+    files = {}
+  end
+  if depth > opts.max_depth then
+    return {}
+  end
+  local fd = vim.uv.fs_scandir(dir)
+  if fd then
+    local name, typ = vim.uv.fs_scandir_next(fd)
+    while name do
+      local path = vim.fs.joinpath(dir, name)
+      if not is_excluded(opts.exclude, path) then
+        if typ == 'file' and vim.tbl_contains(opts.include, 'files') then
+          table.insert(files, path)
+        elseif typ == 'directory' then
+          if vim.tbl_contains(opts.include, 'dirs') then
+            table.insert(files, path)
+          end
+          traverse_directory(opts, path, depth + 1, files)
+        end
+      end
+      name, typ = vim.uv.fs_scandir_next(fd)
+    end
+  end
+  return files
+end
+
+--- Get list of files and directories in the project root
+--- @param opts table|nil Optional configuration table to override default settings
+---   @field root string The project root directory
+---   @field include table A list of types to include in the result, e.g., {'files', 'dirs'}
+---   @field exclude table A list of patterns to exclude from the result, e.g., {'.git', '.hg', '.svn', '.cache', 'node_modules', 'build'}
+---   @field max_depth number The maximum depth to traverse the directory tree (default: 5)
+---   @field follow_links boolean Whether to follow symbolic links (default: false)
+---   @field follow_gitignore boolean Whether to respect the `.gitignore` file (default: true)
+--- @return table A table containing the list of files and directories in the project root
+function M.get_project_files(opts)
```

Comment:
Why is something like this needed, opposed to deferring to something like `git ls-files` ?

Reply:
Fair point. Still not sure if this warrants a separate method given that `vim.fs.dir` already exists and can easily be composed with a root/buffer-dir property/option/function, once that exists.

If it is about the gitignore filtering, I guess we could expose some kind of predicate function, so one can easily filter the iterator returned by `fs.dir`. Would be more flexible as it could be used in other contexts too.

E.g.:

```lua
vim.iter(vim.fs.dir(vim.bo.root):filter(vim.?.in_gitignore) `
```

(But not sure how many projects have a .gitignore without also using git, `git ls-files` does take a path option after all if the root is for some reason in a subdir)